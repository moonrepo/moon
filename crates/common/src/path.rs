use miette::IntoDiagnostic;
pub use relative_path::*;
use rustc_hash::FxHasher;
use starbase_styles::color;
use std::hash::Hasher;
use std::path::Path;

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[inline]
pub fn is_root_level_source<T: AsRef<str>>(source: T) -> bool {
    let source = source.as_ref();
    source.is_empty() || source == "."
}

#[cfg(unix)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

#[cfg(windows)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('/', "\\")
}

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

pub enum RelativeFrom<'path> {
    Project(&'path str),
    Workspace,
}

pub fn expand_to_workspace_relative<P: AsRef<str>>(
    from_format: RelativeFrom,
    path: P,
) -> WorkspaceRelativePathBuf {
    let path = standardize_separators(path.as_ref());

    match from_format {
        RelativeFrom::Project(source) => {
            // Root-level project
            if is_root_level_source(source) {
                WorkspaceRelativePathBuf::from(path)

                // Project-level, prefix with source path
            } else {
                let project_source = standardize_separators(source);

                if let Some(negated_glob) = path.strip_prefix('!') {
                    WorkspaceRelativePathBuf::from(format!("!{project_source}")).join(negated_glob)
                } else {
                    WorkspaceRelativePathBuf::from(project_source).join(path)
                }
            }
        }
        RelativeFrom::Workspace => WorkspaceRelativePathBuf::from(path),
    }
}

#[inline]
pub fn to_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    let path = path.as_ref();

    match path.to_str() {
        Some(p) => Ok(p.to_owned()),
        None => Err(miette::miette!(
            "Path {} contains invalid UTF-8 characters.",
            color::path(path)
        )),
    }
}

#[inline]
pub fn to_virtual_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    Ok(standardize_separators(to_string(path)?))
}

#[inline]
pub fn to_relative_virtual_string<F: AsRef<Path>, T: AsRef<Path>>(
    from: F,
    to: T,
) -> miette::Result<String> {
    let value = from
        .as_ref()
        .relative_to(to.as_ref())
        .into_diagnostic()?
        .to_string();

    Ok(if value.is_empty() { ".".into() } else { value })
}

#[inline]
pub fn exe_name<T: AsRef<str>>(name: T) -> String {
    #[cfg(windows)]
    {
        format!("{}.exe", name.as_ref())
    }

    #[cfg(not(windows))]
    {
        name.as_ref().into()
    }
}

/// Encode a value (typically an identifier) by removing invalid characters for use
/// within a file name.
pub fn encode_component(value: impl AsRef<str>) -> String {
    let mut output = String::new();

    // Handle supported characters from `Id`
    for ch in value.as_ref().chars() {
        match ch {
            '@' | '*' => {
                // Skip these
            }
            '/' | ':' => {
                output.push('-');
            }
            _ => {
                output.push(ch);
            }
        }
    }

    output.trim_matches(['-', '.']).to_owned()
}

/// Hash a value that may contain special characters into a valid file name.
pub fn hash_component(value: impl AsRef<str>) -> String {
    let mut hasher = FxHasher::default();
    hasher.write(value.as_ref().as_bytes());

    format!("{}", hasher.finish())
}
