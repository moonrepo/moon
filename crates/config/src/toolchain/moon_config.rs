use crate::config_struct;
use schematic::{Config, validate};

config_struct!(
    /// Configures how and where updates will be received.
    #[derive(Config)]
    pub struct MoonConfig {
        /// A secure URL to lookup the latest version.
        #[setting(validate = validate::url_secure, default = "https://launch.moonrepo.app/versions/cli/current")]
        pub manifest_url: String,

        /// A secure URL for downloading the moon binary.
        #[setting(validate = validate::url_secure, default = "https://github.com/moonrepo/moon/releases/latest/download")]
        pub download_url: String,
    }
);
