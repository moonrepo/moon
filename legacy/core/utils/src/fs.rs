use crate::time::{is_stale, to_millis};
use moon_common::consts::CONFIG_DIRNAME;
use starbase_utils::fs;
use std::path::{Path, PathBuf};
use std::time::Duration;

pub mod temp {
    use super::*;

    pub fn get_dir() -> PathBuf {
        crate::get_workspace_root()
            .join(CONFIG_DIRNAME)
            .join("cache")
            .join("temp")
    }

    pub fn get_file(source: &str, ext: &str) -> PathBuf {
        get_dir().join(format!("{}.{}", crate::hash(source), ext))
    }

    pub fn read<P: AsRef<Path>>(path: P) -> miette::Result<Option<String>> {
        let file = path.as_ref();

        if !file.exists() {
            return Ok(None);
        }

        if let Ok(metadata) = file.metadata()
            && let Ok(filetime) = metadata.created()
        {
            // Temp files only last for 4 hours (half a workday)
            if is_stale(to_millis(filetime), Duration::from_secs(60 * 60 * 4)) {
                fs::remove_file(file)?;

                return Ok(None);
            }
        }

        Ok(Some(fs::read_file(file)?))
    }

    pub fn write<P: AsRef<Path>, D: AsRef<str>>(path: P, data: D) -> miette::Result<()> {
        let path = path.as_ref();

        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::write_file(path, data.as_ref())?;

        Ok(())
    }
}
