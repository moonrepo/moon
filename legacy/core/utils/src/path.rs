use miette::IntoDiagnostic;
use moon_common::path::clean_components;
use relative_path::PathExt;
use std::path::{Path, PathBuf};

pub use pathdiff::diff_paths as relative_from;

/// If a file starts with "/", expand from the workspace root, otherwise the project root.
#[inline]
pub fn expand_to_workspace_relative<F, P>(file: F, workspace_root: P, project_root: P) -> PathBuf
where
    F: AsRef<str>,
    P: AsRef<Path>,
{
    let file = file.as_ref();
    let workspace_root = workspace_root.as_ref();
    let project_root = project_root.as_ref();

    if let Some(ws_rel_file) = file.strip_prefix('/') {
        return PathBuf::from(normalize_separators(ws_rel_file));
    }

    let project_source = project_root.strip_prefix(workspace_root).unwrap();

    if let Some(negative_glob) = file.strip_prefix('!') {
        return PathBuf::from(format!("!{}", project_source.to_string_lossy()))
            .join(normalize_separators(negative_glob));
    }

    project_source.join(normalize_separators(file))
}

#[inline]
pub fn normalize<T: AsRef<Path>>(path: T) -> PathBuf {
    clean_components(path.as_ref())
}

#[cfg(not(windows))]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

#[cfg(windows)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('/', "\\")
}

#[inline]
pub fn replace_home_dir<T: AsRef<str>>(value: T) -> String {
    let value = value.as_ref();

    if let Some(home_dir) = starbase_utils::dirs::home_dir() {
        let home_dir_str = home_dir.to_str().unwrap_or_default();

        // Replace both forward and backward slashes
        return value
            .replace(home_dir_str, "~")
            .replace(&standardize_separators(home_dir_str), "~");
    }

    value.to_owned()
}

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

#[inline]
pub fn to_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    let mut path = path.as_ref();

    // Avoid UNC paths as they cause lots of issues
    if cfg!(windows) {
        path = dunce::simplified(path);
    }

    match path.to_str() {
        Some(p) => Ok(p.to_owned()),
        None => Err(miette::miette!(
            "Path {} contains invalid UTF-8 characters.",
            path.display()
        )),
    }
}

#[inline]
pub fn to_virtual_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    Ok(standardize_separators(to_string(path)?))
}

#[inline]
pub fn to_relative_virtual_string<F: AsRef<Path>, T: AsRef<Path>>(
    from: F,
    to: T,
) -> miette::Result<String> {
    let value = from
        .as_ref()
        .relative_to(to.as_ref())
        .into_diagnostic()?
        .to_string();

    Ok(if value.is_empty() { ".".into() } else { value })
}
