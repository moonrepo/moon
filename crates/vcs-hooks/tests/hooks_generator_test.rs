use moon_config::VcsConfig;
use moon_vcs::{BoxedVcs, Git};
use moon_vcs_hooks::HooksGenerator;
use rustc_hash::FxHashMap;
use starbase_sandbox::{assert_snapshot, create_empty_sandbox};
use std::fs;
use std::path::Path;

fn load_git(root: &Path) -> BoxedVcs {
    Box::new(Git::load(root, "master", &[]).unwrap())
}

fn create_config() -> VcsConfig {
    VcsConfig {
        hooks: FxHashMap::from_iter([
            (
                "pre-commit".into(),
                vec!["moon run :lint".into(), "some-command".into()],
            ),
            ("post-push".into(), vec!["moon check --all".into()]),
        ]),
        ..VcsConfig::default()
    }
}

async fn run_generator(root: &Path) {
    HooksGenerator::new(&load_git(root), &create_config(), root)
        .generate()
        .await
        .unwrap();
}

async fn clean_generator(root: &Path) {
    HooksGenerator::new(&load_git(root), &create_config(), root)
        .cleanup()
        .await
        .unwrap();
}

#[tokio::test]
async fn doesnt_generate_when_no_hooks() {
    let sandbox = create_empty_sandbox();
    sandbox.enable_git();

    HooksGenerator::new(
        &load_git(sandbox.path()),
        &VcsConfig::default(),
        sandbox.path(),
    )
    .generate()
    .await
    .unwrap();

    assert!(!sandbox.path().join(".moon/hooks").exists());
}

#[tokio::test]
async fn doesnt_generate_when_no_commands() {
    let sandbox = create_empty_sandbox();
    sandbox.enable_git();

    HooksGenerator::new(
        &load_git(sandbox.path()),
        &VcsConfig {
            hooks: FxHashMap::from_iter([
                ("pre-commit".into(), vec![]),
                ("post-push".into(), vec![]),
            ]),
            ..VcsConfig::default()
        },
        sandbox.path(),
    )
    .generate()
    .await
    .unwrap();

    assert!(!sandbox.path().join(".moon/hooks").exists());
}

#[tokio::test]
async fn cleans_up_hooks() {
    let sandbox = create_empty_sandbox();
    sandbox.enable_git();

    run_generator(sandbox.path()).await;

    let pre_commit = sandbox.path().join(".git/hooks/pre-commit");
    let post_push = sandbox.path().join(".git/hooks/post-push");
    let local_hooks = sandbox.path().join(".moon/hooks");

    assert!(pre_commit.exists());
    assert!(post_push.exists());
    assert!(local_hooks.exists());

    clean_generator(sandbox.path()).await;

    assert!(!pre_commit.exists());
    assert!(!post_push.exists());
    assert!(!local_hooks.exists());
}

#[cfg(unix)]
mod unix {
    use super::*;
    use std::path::PathBuf;

    #[tokio::test]
    async fn doesnt_create_outside_repo_root() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();
        sandbox.run_git(|git| {
            git.args(["config", "--local", "core.hooksPath", "/tmp/git_hooks"]);
        });

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".git/hooks/pre-commit");
        let post_push = sandbox.path().join(".git/hooks/post-push");

        assert!(pre_commit.exists());
        assert!(post_push.exists());
        assert!(!PathBuf::from("/tmp/git_hooks").exists());
    }

    #[tokio::test]
    async fn creates_local_hook_files() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".moon/hooks/pre-commit.sh");
        let post_push = sandbox.path().join(".moon/hooks/post-push.sh");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert_snapshot!(fs::read_to_string(pre_commit).unwrap());
        assert_snapshot!(fs::read_to_string(post_push).unwrap());

        let pre_commit = sandbox.path().join(".git/hooks/pre-commit");
        let post_push = sandbox.path().join(".git/hooks/post-push");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert!(
            fs::read_to_string(pre_commit)
                .unwrap()
                .contains("./.moon/hooks/pre-commit.sh $1 $2 $3")
        );
        assert!(
            fs::read_to_string(post_push)
                .unwrap()
                .contains("./.moon/hooks/post-push.sh $1 $2 $3")
        );
    }

    #[tokio::test]
    async fn creates_hook_files_with_trailing_newline() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".moon/hooks/pre-commit.sh");

        assert!(pre_commit.exists());
        assert_eq!(
            fs::read_to_string(pre_commit).unwrap().chars().last(),
            Some('\n')
        )
    }

    #[tokio::test]
    async fn links_git_hooks() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".git/hooks/pre-commit");
        let post_push = sandbox.path().join(".git/hooks/post-push");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert_snapshot!(fs::read_to_string(pre_commit).unwrap());
        assert_snapshot!(fs::read_to_string(post_push).unwrap());
    }

    #[tokio::test]
    async fn doesnt_links_git_hooks_if_no_root() {
        let sandbox = create_empty_sandbox();

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".git/hooks/pre-commit");
        let post_push = sandbox.path().join(".git/hooks/post-push");

        assert!(!pre_commit.exists());
        assert!(!post_push.exists());
    }

    #[tokio::test]
    async fn supports_git_worktrees() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        sandbox.run_git(|cmd| {
            cmd.args(["worktree", "add", "tree"]);
        });

        run_generator(&sandbox.path().join("tree")).await;

        let pre_commit = sandbox.path().join("tree/.moon/hooks/pre-commit.sh");
        let post_push = sandbox.path().join("tree/.moon/hooks/post-push.sh");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        let pre_commit = sandbox.path().join(".git/worktrees/tree/hooks/pre-commit");
        let post_push = sandbox.path().join(".git/worktrees/tree/hooks/post-push");

        assert!(pre_commit.exists());
        assert!(post_push.exists());
    }
}

#[cfg(windows)]
mod windows {
    use super::*;
    use moon_config::VcsHookFormat;

    // Standardize snapshots across machines with different powershell versions
    fn clean_powershell(content: String) -> String {
        content.replace("powershell", "pwsh")
    }

    #[tokio::test]
    async fn creates_local_hook_files() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".moon/hooks/pre-commit.ps1");
        let post_push = sandbox.path().join(".moon/hooks/post-push.ps1");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert_snapshot!(clean_powershell(fs::read_to_string(pre_commit).unwrap()));
        assert_snapshot!(clean_powershell(fs::read_to_string(post_push).unwrap()));

        let pre_commit = sandbox.path().join(".git/hooks/pre-commit");
        let post_push = sandbox.path().join(".git/hooks/post-push");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert!(
            fs::read_to_string(pre_commit)
                .unwrap()
                .contains(".\\.moon\\hooks\\pre-commit.ps1")
        );
        assert!(
            fs::read_to_string(post_push)
                .unwrap()
                .contains(".\\.moon\\hooks\\post-push.ps1")
        );
    }

    #[tokio::test]
    async fn creates_local_hook_files_as_bash() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        let mut config = create_config();
        config.hook_format = VcsHookFormat::Bash;

        HooksGenerator::new(&load_git(sandbox.path()), &config, sandbox.path())
            .generate()
            .await
            .unwrap();

        let pre_commit = sandbox.path().join(".moon/hooks/pre-commit.sh");
        let post_push = sandbox.path().join(".moon/hooks/post-push.sh");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert_snapshot!(clean_powershell(fs::read_to_string(pre_commit).unwrap()));
        assert_snapshot!(clean_powershell(fs::read_to_string(post_push).unwrap()));

        let pre_commit = sandbox.path().join(".git/hooks/pre-commit");
        let post_push = sandbox.path().join(".git/hooks/post-push");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert!(
            fs::read_to_string(pre_commit)
                .unwrap()
                .contains("./.moon/hooks/pre-commit.sh $1 $2 $3")
        );
        assert!(
            fs::read_to_string(post_push)
                .unwrap()
                .contains("./.moon/hooks/post-push.sh $1 $2 $3")
        );
    }

    #[tokio::test]
    async fn links_git_hooks() {
        let sandbox = create_empty_sandbox();
        sandbox.enable_git();

        run_generator(sandbox.path()).await;

        let pre_commit = sandbox.path().join(".git/hooks/pre-commit");
        let post_push = sandbox.path().join(".git/hooks/post-push");

        assert!(pre_commit.exists());
        assert!(post_push.exists());

        assert_snapshot!(clean_powershell(fs::read_to_string(pre_commit).unwrap()));
        assert_snapshot!(clean_powershell(fs::read_to_string(post_push).unwrap()));
    }
}
