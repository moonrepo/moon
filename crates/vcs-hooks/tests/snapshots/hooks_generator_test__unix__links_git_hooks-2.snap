---
source: crates/vcs-hooks/tests/hooks_generator_test.rs
expression: "fs::read_to_string(post_push).unwrap()"
---
#!/usr/bin/env bash
set -eo pipefail

for (( i=0; i <= $#; i++ )); do
  declare -x "ARG$i"="${!i}"
done

./.moon/hooks/post-push.sh $1 $2 $3 $4 $5
